### ISSUE TITLE
API: Native Library Resolve Event

### ISSUE THREAD

**Issue by swaroop-sridhar** [dotnet/corefx#32850]
This proposal adds a Native library resolving event, to be raised when the runtime cannot resolve a native library load.

# Proposed API
```C#
namespace System.Runtime.Loader
{
    public abstract class AssemblyLoadContext
    {
        /// Event handler for resolving native libraries
        /// Inputs: Invoking assembly, and library name to resolve
        /// Returns: A handle to the loaded native library
        public event Func<Assembly, string, IntPtr> ResolvingUnmanagedDll;
    }
}
```
# Rationale 
In the case of Loading assemblies, the runtime provides the application to customize the load at various phases:
- Before running the default load logic via [`AssemblyLoadContext.Load()`](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs#L273) method overrides
- After running the default load logic via the [`AssemblyLoadContext.Resolving`](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs#L524) event handlers.

This proposal creates a matching facility for unmanaged libraries
- Before default DllImport logic: The existing [`AssemblyLoadContext.LoadUnmanagedDll()`](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs#L405) and separately proposed [`DllImportResolver`](https://github.com/dotnet/corefx/issues/32015) callbacks
- After default DllImport logic: The new `AssemblyLoadContext.ResolvingUnmanagedDll` event.

In particular, the `NativeLibraryResolve` event is expected to provide flexibility for running custom native library resolution logic from plugins. Hence the motivation to have the event per AssemblyLoadContext, rather than globally.

### DllImport sequence

DllImport load library works in the following order, stop at any step the library is successfully loaded.
* If the invoking-assembly has a `DllImportResolver` callback registered, invoke it.
* If the invoking-assembly is not in the default load context, call `AssemblyLoadContext.LoadUnmanagedDll()`
* Run the default load logic, try loading from:
    * AppDomain cache
    * NATIVE_DLL_SEARCH_DIRECTORIES
    * Invoking-assembly directory, System32, etc. based on `DllImportSearchPaths`
* Raise the ResolvingUnmanagedDll event

# Discussion
### Alternate notation
In this proposal `ResolvingUnmanagedDll` uses `Func` notation in order to be consistent with the 
[Loading/Unloading events](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs#L512-L513) in `AssemblyLoadContext`.

An alternative is to use `EventHandler` style recommended by these [guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/event), and similar to certain other [events](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs#L526-L537) in the same class.

```C#
namespace System
{
    public delegate IntPtr UnmanagedDllResolveEventHandler(object sender, ResolveEventArgs args);
}

namespace System.Runtime.Loader
{
    public abstract class AssemblyLoadContext
    {
        public event UnmanagedDllResolveEventHandler ResolvingUnmanagedDll;
    }
}
```

### Related Topics
#32015 Native Library Loader API

**Comment by swaroop-sridhar** [comment]
CC: @vitek-karas @jeffschwMSFT @AaronRobinsonMSFT @jkotas

**Comment by jkotas** [comment]
`ResolveEventHandler` returns `Assembly`. This should have a different type.

**Comment by swaroop-sridhar** [comment]
Thanks @jkotas, I've fixed it now.

**Comment by jkotas** [comment]
The existing events on AssemblyLoadContext use Action/Func. Do we want to follow the same style?

**Comment by swaroop-sridhar** [comment]
In `AssemblyLoadContext`, the [Loading/Unloading events](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs#L524-L525) use the `Func` notation, while [others](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs#L539-L565) use `EventHandler` style.

I wrote in event in the EventHandler style following these [guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/event).
I'm happy to write this event in the Func style, in fact, I'd prefer it for this case.

**Comment by vitek-karas** [comment]
I'm worried about naming consistency. On managed side we have `Load`/`Resolving`. On native we have `LoadUnmanagedDll`/`NativeLibraryResolve`. To be 100% consistent we should call it `ResolvingUnmanagedDll`.

**Comment by vitek-karas** [comment]
I think it's important to mention that there are 2 reasons to add this:
- Plugins - which don't want to register callbacks for every assembly in the plugin as that creates problems with multiple callbacks support.
- Anybody (Plugins included) - as this will be the only way (excluding callbacks) to provide custom resolution of native libraries in the Default context.

**Comment by vitek-karas** [comment]
I also think it would be a good idea to describe the proposed ordering of the various hooks. It starts with a call to a `DllImport` method - what are the high-level steps and where/when are the various hooks called.

**Comment by AaronRobinsonMSFT** [comment]
>what are the high-level steps and where/when are the various hooks called.

Following on to @vitek-karas's comment, the official guidance for [`DllImport`](https://docs.microsoft.com/en-us/cpp/dotnet/calling-native-functions-from-managed-code?view=vs-2017) should be updated with these newdetails.

**Comment by swaroop-sridhar** [comment]
I've updated the naming, and added a section on DllImport order.

**Comment by swaroop-sridhar** [comment]
> the official guidance for [`DllImport`](https://docs.microsoft.com/en-us/cpp/dotnet/calling-native-functions-from-managed-code?view=vs-2017) should be updated with these newdetails.
Sure, once the feature is checked in/released, we should update it.

**Comment by jeffschwMSFT** [comment]
Closed with #34668
